contracts/PrepaidCardMarketV2.sol

```diff
@@ -0,0 +1,359 @@
+      token,
+      faceValue
+    ); // in Wei
+
+    address issuer = skus[sku].issuer;
+    address issuerSafe = skus[sku].issuerSafe;
+
+    require(
+      balance[issuerSafe][token] >= priceToCreatePrepaidCard,
+      "Not enough balance"
+    );
+
+    uint256[] memory issuingTokenAmounts = new uint256[](1);
+    uint256[] memory spendAmounts = new uint256[](1);
+
+    issuingTokenAmounts[0] = priceToCreatePrepaidCard;
+    spendAmounts[0] = faceValue;
+
+    balance[issuerSafe][token] -= priceToCreatePrepaidCard;
+
+    IERC677(token).safeTransferAndCall(
+      prepaidCardManagerAddress,
+      priceToCreatePrepaidCard,
+      abi.encode(
+        customer,
+        issuingTokenAmounts,
+        spendAmounts,
+        customizationDID,
+        address(0), // marketAddress - we don't need it in this case
+        issuer,
+        issuerSafe
+      )
+    );
+
+    emit PrepaidCardProvisioned(customer, sku);
+
+    return true;
+  }
+
+  function getQuantity(bytes32 sku) public view returns (uint256) {
+    PrepaidCardManager prepaidCardManager = PrepaidCardManager(
+      prepaidCardManagerAddress
+    );
+
+    address token = skus[sku].issuingToken;
+    uint256 faceValue = skus[sku].faceValue;
+    address issuerSafe = skus[sku].issuerSafe;
+
+    uint256 price = prepaidCardManager.priceForFaceValue(token, faceValue);
+
+    return balance[issuerSafe][token] / price;
+  }
+
+  function setAsk(
+    address issuer,
+    bytes32 sku,
+    uint256 askPrice
+  ) external onlyHandlers returns (bool) {
+    require(skus[sku].issuer != address(0), "Non-existent SKU");
+    require(skus[sku].issuer == issuer, "SKU not owned by issuer");
+
+    asks[sku] = askPrice;
+
+    emit AskSet(issuer, skus[sku].issuingToken, sku, askPrice);
+    return true;
+  }
+
+  function addSKU(
+    address issuerSafe,
+    uint256 faceValue,
+    string memory customizationDID,
+    address token
+  ) external onlyHandlers returns (bool) {
+    require(faceValue > 0, "Face value must be greater than 0");
+    require(token != address(0), "Token address must be set");
+
+    address _issuer = issuers[issuerSafe];
+    require(_issuer != address(0), "Issuer has no balance");
+
+    bytes32 sku = getSKU(_issuer, token, faceValue, customizationDID);
+
+    require(skus[sku].issuer == address(0), "SKU already exists");
+
+    skus[sku] = SKU({
+      issuerSafe: issuerSafe,
+      issuer: _issuer,
+      issuingToken: token,
+      faceValue: faceValue,
+      customizationDID: customizationDID
+    });
+
+    emit SkuAdded(_issuer, token, faceValue, customizationDID, sku);
+
+    return true;
+  }
+
+  function getSKU(
+    address issuer,
+    address token,
+    uint256 faceValue,
+    string memory customizationDID
+  ) public pure override returns (bytes32) {
+    return
+      keccak256(abi.encodePacked(issuer, token, faceValue, customizationDID));
+  }
+
+  function withdrawTokens(uint256 amount, address token) external {
```

Comment:
I have had a thought that I'm not sure of the answer to:

What are the implications if an issue safe changes owners.

* Is it still possible to withdraw tokens?
* are there other issues for the other functionality of the contract?

Not sure of the answer to either question